internal/generator/generator_util.go

```diff
@@ -0,0 +1,200 @@
+	byteContent, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	strContent := string(byteContent)
+
+	// Remove comments
+	strContent = singleLineCommentExtracter.ReplaceAllString(strContent, "")
+	strContent = multiLineCommentExtracter.ReplaceAllString(strContent, "")
+	strContent = strings.ReplaceAll(strContent, "\r\n", "")
+	strContent = strings.ReplaceAll(strContent, "\n", "")
+
+	// Extract directives definition code blocks, each code block contains a list of directives definition
+	blocks := directivesDefineBlockExtracter.FindAllStringSubmatch(strContent, -1)
+
+	for _, block := range blocks {
+		// Extract directives and their attributes in the code block, the first dimension of subBlocks
+		// is index of directive, the second dimension is index of attributes
+		subBlocks := singleDirectiveExtracter.FindAllStringSubmatch(block[2], -1)
+
+		// Iterate through every directive
+		for _, attributes := range subBlocks {
+			// Extract attributes from the directive
+			directiveName := strings.TrimSpace(attributes[1])
+			directiveMask := strings.Split(attributes[2], "|")
+
+			// Transfer C-style mask to go style
+			for idx, ngxVarName := range directiveMask {
+				goVarName, found := ngxVarNameToGo[strings.TrimSpace(ngxVarName)]
+				if !found {
+					return nil, fmt.Errorf("parsing directive %s, bitmask %s in source code not found in crossplane", directiveName, ngxVarName)
+				}
+				directiveMask[idx] = goVarName
+			}
+
+			directive2Masks[directiveName] = append(directive2Masks[directiveName], directiveMask)
+		}
+	}
+	return directive2Masks, nil
+}
+
+//nolint:nonamedreturns
+func getMasksFromPath(path string) (directive2Masks map[string][]mask, err error) {
+	directive2Masks = make(map[string][]mask, 0)
+
+	err = filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Check if the entry is a C/C++ file
+		// Some dynamic modules are written in C++, like otel
+		if d.IsDir() {
+			return nil
+		}
+
+		if !(strings.HasSuffix(path, ".c") || strings.HasSuffix(path, ".cpp")) {
+			return nil
+		}
+
+		directive2MasksInFile, err := masksFromFile(path)
+		if err != nil {
+			return err
+		}
+
+		for directive, masksInFile := range directive2MasksInFile {
+			directive2Masks[directive] = append(directive2Masks[directive], masksInFile...)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(directive2Masks) == 0 {
+		return nil, errors.New("can't find any directives in the directory and subdirectories, please check the path")
+	}
+
+	return directive2Masks, nil
+}
+
+func genFromSrcCode(codePath string, mapVariableName string, matchFnName string, writer io.Writer) error {
+	directive2Masks, err := getMasksFromPath(codePath)
+	if err != nil {
+		return err
+	}
+
+	err = supportFileTmpl.Execute(writer, supportFileTmplStruct{
+		Directive2Masks: directive2Masks,
+		MapVariableName: mapVariableName,
+		MatchFnName:     matchFnName,
```

Comment:
Are `MapVariableName` and `MatchFnName` expected to change? Just curious why you made them template variables

Reply:
Yeah there will be another generator directly fetch code from some github repos and these names will change accordingly. For example, the generator fetches Lua source code and the generated `MapVariableName` will be `moduleLuaDirectives` and the `MatchFnName` will be `LuaDirectivesMatchFn`. That tool aims for auto update current existing support files(`analyze_xxx_directives.go`)

Reply:
Got it, thanks for the explanation!